client/base_client.go

```diff
@@ -12,3 +13,31 @@ type BaseClient interface {
 	SendRequest(method string, rawURL string, data url.Values,
 		headers map[string]interface{}) (*http.Response, error)
 }
+
+// BaseClientWithCtx is an extension of BaseClient with the ability to associate a contex with
+// the request
+type BaseClientWithCtx interface {
+	BaseClient
+	SendRequestWithCtx(ctx context.Context, method string, rawURL string, data url.Values,
+		headers map[string]interface{}) (*http.Response, error)
+}
+
+type wrapperClient struct {
+	// embed the BaseClient so the functions remain accessible
+	BaseClient
+}
+
+func (w wrapperClient) SendRequestWithCtx(ctx context.Context, method string, rawURL string, data url.Values,
+	headers map[string]interface{}) (*http.Response, error) {
+	return w.SendRequest(method, rawURL, data, headers)
+}
+
+// wrapBaseClientWithNoopCtx "upgrades" a BaseClient to BaseClientWithCtx so that requests can be
+// send with a request context.
+func wrapBaseClientWithNoopCtx(c BaseClient) BaseClientWithCtx {
```

Comment:
nit: not really a `Noop` context since the noop gets used when calling `SendRequest` even when not wrapping the base client.